fastavro/_read_py.py

```diff
@@ -478,34 +479,68 @@ def skip_sync(fo, sync_marker):
         raise ValueError('expected sync marker not found')
 
 
+class ReaderBase(object):
+    def __init__(self):
+        self._position = 0
+
+    def position(self):
+        return self._position
+
+
+class MemoryReader(ReaderBase):
+    def __init__(self, data):
+        super(MemoryReader, self).__init__()
+        self.data = data
+
+    def read(self, size):
+        result = self.data[self._position: self._position + size]
+        self._position += size
+        return result
+
+    def __len__(self):
+        return len(self.data)
+
+
+class FileObjectReader(ReaderBase):
```

Comment:
I know the cython side needed a special `Reader` object in the past because the underlying cython type couldn't be used in the same way as the python type. Is the need here on the python side basically just for being able to call `len()` (as it seems the `MemoryIO` object doesn't have a way to do that)?

Reply:
yea, I didn't have a well-thought-through story here, thanks for calling out!

in the last two commits in the latest batch (https://github.com/tebeka/fastavro/pull/208/commits/2cfe013495aa41ea02feef50e3254c55067ceb18 and https://github.com/tebeka/fastavro/pull/208/commits/ca99dcb011803ed0627d03d86d9b08c9fb16dfdf), I took a stricter eye to which of the abstractions {`ReaderBase`, `MemoryReader`, `FileObjectReader`} were needed in each of {`.pyx`, `.py`}.

afaict, none of them are necessary!

The stream-like things we deal with are:
- "files" (which can be actual file-descriptors, or in-memory buffers simulating files in tests), and
- memory-buffers representing avro blocks.

The former need `read` and `tell` methods (`tell` is used for inferring block size after reading it), and the latter need only `read`. File-descriptors and `MemoryIO`s both support both.

I'm trying to run some of the benchmarks now to see whether there's a noticeable difference between the versions with or without these abstractions.

Reply:
Great. That should reduce some complexity. We probably should have added a test if the original cython ones were needed, but it's possible it was done for some micro-optimizations. However, I think the reduced complexity and more similar to the python implementation are more beneficial.